src/server/actions/invite.actions.ts

```diff
@@ -0,0 +1,192 @@
+ *     marking the invitation as claimed.
+ */
+'use server'
+
+import { z } from 'zod'
+import { clerkClient } from '@clerk/nextjs/server'
+import { db } from '@/database'
+import { ActionFailure, isClerkApiError } from '@/lib/errors'
+import { findOrCreateOrgMembership } from '@/server/mutations' // DB only
+import logger from '@/lib/logger'
+import { findClerkOrganization } from '@/server/clerk'
+import { onUserAcceptInvite } from '@/server/events'
+import { anonAction, userAction, actionContext } from '@/server/actions/wrappers'
+
+// Schema for onCreateAccountAction input
+const CreateAccountSchema = z.object({
+    inviteId: z.string(),
+    email: z.string().email(), // Added email to ensure it matches invite
+    form: z.object({
+        firstName: z.string().nonempty(),
+        lastName: z.string().nonempty(),
+        password: z.string().min(8),
+    }),
+})
+
+export const onCreateAccountAction = anonAction(async ({ inviteId, email, form }) => {
+    const pendingUser = await db
+        .selectFrom('pendingUser')
+        .selectAll('pendingUser')
+        .where('id', '=', inviteId)
+        .where('email', '=', email) // Verify email matches the invite
+        .where('claimedByUserId', 'is', null)
+        .executeTakeFirst()
+
+    if (!pendingUser) {
+        throw new ActionFailure({ form: 'Invalid or already claimed invitation.' })
+    }
+
+    try {
+        const client = await clerkClient()
+        const existingClerkUsers = await client.users.getUserList({ emailAddress: [email] })
+        if (existingClerkUsers.data.length > 0) {
+            // This case should ideally be handled by InvitationHandler routing to existing user flow.
+            // If an existing user somehow reaches here, it's an anomaly.
+            throw new ActionFailure({
+                email: 'An account with this email already exists. Please sign in to accept the invitation.',
+            })
+        }
+
+        const clerkUser = await client.users.createUser({
+            emailAddress: [email],
+            password: form.password,
+            firstName: form.firstName,
+            lastName: form.lastName,
+            // publicMetadata for roles will be set by onPendingUserLoginAction
+        })
+        return clerkUser.id // Return Clerk User ID
+    } catch (error: unknown) {
+        if (isClerkApiError(error)) {
+            const clerkError = error.errors[0]
+            logger.warn({ message: 'Clerk API error during user creation', clerkError, email })
+            throw new ActionFailure({ form: clerkError.longMessage || clerkError.message })
+        }
+        logger.error({ message: 'Non-Clerk error during user creation', error, email })
+        throw new ActionFailure({ form: 'Could not create account due to an unexpected error.' })
+    }
+}, CreateAccountSchema)
+
+const ClaimInviteSchema = z.object({
+    inviteId: z.string(),
+})
+
+export const claimInviteAction = userAction(async ({ inviteId }) => {
```

Comment:
this is called after the MFA setup is complete, correct?  If that's the case then it seems like all it should do is look update any invitations that are for a lowercase email matching the current user's email to be claimed.

All the other checks you have here should be front-loaded so the user hits them before going through the acceptance flow

Reply:
done